### PR TITLE
Added tracing to the relay client and server.

### PR DESCRIPTION
**Pull request by hauke76**
This adds trace support to our relay server and client to be able to better debug and monitor its performance.

Traces are logged when they are enabled through the TRSID cookie.

The generated traces (see image below) look at times strange. My working hypothesis is that the relay server sends its reply to the user client (see `server.userClientRequest()`) before actually finishing the async receiving of the response in `server.serverResponse()`.

image

### CODE REVIEW

File: src/go/cmd/http-relay-client/main.go [hunk 1]
```diff
@@ -152,35 +163,31 @@ func marshalHeader(h *http.Header) []*pb.HttpHeader {
...
+		header.Add(*h.Name, *h.Value)
+	}
+}
+
+func createBackendRequest(breq *pb.HttpRequest) (*http.Request, error) {
 	id := *breq.Id
 	targetUrl, err := url.Parse(*breq.Url)
-	if err != nil {
-		return nil, nil, err
-	}
```
Comment (ensonic) [inline src/go/cmd/http-relay-client/main.go:165]: why this change?

File: src/go/cmd/http-relay-client/main.go [hunk 2]
```diff
@@ -190,10 +197,31 @@ func makeBackendRequest(local *http.Client, breq *pb.HttpRequest) (*pb.HttpRespo
 		log.Printf("%s", dump)
 	}
 
+	return req, nil
+}
+
+// This function builds and executes a http.Request from the proto request we received from the web-client.
```
Comment (ensonic) [inline src/go/cmd/http-relay-client/main.go:203]: "user-client" or maybe end-user client. here and below

File: src/go/cmd/http-relay-client/main.go [hunk 3]
```diff
@@ -449,10 +511,16 @@ func handleRequest(remote *http.Client, local *http.Client, req *pb.HttpRequest)
...
 					resp.BackendDurationMs = proto.Int64(duration.Milliseconds())
 					// see makeBackendRequest()
-					urlPath := strings.TrimPrefix(*req.Url, "http://invalid")
+					urlPath := strings.TrimPrefix(*pbreq.Url, "http://invalid")
 					log.Printf("[%s] Backend request duration: %.3fs (for %s)", *resp.Id, duration.Seconds(), urlPath)
+				} else {
+					// Q(hauke): When are we ending up in this branch?
+					// What are the semantics and why are we not setting a request duration?
+					// Even in a streaming case I would expect a duration which represents the
+					// processing time of the last item.
```
Comment (ensonic) [inline src/go/cmd/http-relay-client/main.go:520]: we could track each backend request and report the progress. let try as a followup

File: src/go/cmd/http-relay-client/main.go [hunk 4]
```diff
@@ -449,10 +511,16 @@ func handleRequest(remote *http.Client, local *http.Client, req *pb.HttpRequest)
...
 				}
-				return postResponse(remote, resp)
+				err = postResponse(remote, resp)
+				return err
```
Comment (ensonic) [inline src/go/cmd/http-relay-client/main.go:523]: revert?

File: src/go/cmd/http-relay-client/main.go [hunk 5]
```diff
@@ -563,7 +645,7 @@ func main() {
 			}
 		}
 
-		transport = h2transport
+		transport = &ochttp.Transport{Base: h2transport}
```
Comment (ensonic) [inline src/go/cmd/http-relay-client/main.go:648]: move `&ochttp.Transport{Base: ` to where we set the transport: `Transport: transport,`

File: src/go/cmd/http-relay-server/broker.go [hunk 6]
```diff
@@ -127,6 +132,7 @@ func (r *broker) RelayRequest(server string, request *pb.HttpRequest) (<-chan *p
 		r.req[server] = make(chan *pb.HttpRequest)
 	}
 	if r.resp[id] != nil {
+		r.m.Unlock()
```
Comment (ensonic) [inline src/go/cmd/http-relay-server/broker.go:135]: wow, could you send this as a separate PR (you can send a PR right from the github webpage).

File: src/go/cmd/http-relay-server/broker.go [hunk 7]
```diff
@@ -212,7 +221,7 @@ func (r *broker) PutRequestStream(id string, data []byte) bool {
 	return true
 }
 
-// SendResponse delivers the HttpResponse to the client handler that created the
+// SendResponse delivers the HttpResponse to the web-client handler that created the
```
Comment (ensonic) [inline src/go/cmd/http-relay-server/broker.go:224]: again, user-client?

**Review by ensonic — COMMENTED** [review]
Thanks!

Do we need a way to configure the sampling rate?

File: src/go/cmd/http-relay-server/broker.go (hunk 7 above)
Comment (hauke76) [reply src/go/cmd/http-relay-server/broker.go:224]: Found a few more places and fixed them too.

File: src/go/cmd/http-relay-server/broker.go (hunk 6 above)
Comment (hauke76) [reply src/go/cmd/http-relay-server/broker.go]: Fixed separately in #173

**Comment by hauke76** [comment]
According to Christian, we don't need to specify the sample rate as it will be propagated. Another question which I would like to address in a follow up PR is whether we enable tracing by default. Christian added it as a safety guard to prevent excessive cost generation.

**Review by ensonic — APPROVED** [review]
thanks!
